Allow paths with /usr/sbin and /usr/bin as equivalent

keszybz opened the pull request:
After https://fedoraproject.org/wiki/Changes/Unify_bin_and_sbin, various rpms change paths from /usr/sbin to /usr/bin. If we're built in an environment where %_sbindir==%_bindir, allow paths to differ in the prefix.

Tested by reverting the patch to move paths in iptables-legacy back to /usr/sbin [1]. The subsequent rpm installs fine both when /usr/sbin is a symlink (new installs) and a directory (upgrades).

This approach is bit of a hack, but it also seems quite safe, because it doesn't change the stored data in any way, it just suppresses two error checks.

[1] https://src.fedoraproject.org/rpms/iptables/c/43696a4be35ecb64b596d1db368cb736a91316e8

lnykryn commented:
Chmmm, I guess this fixes the immediate problem, so maybe we should just merge it.
But honestly I would be glad if simple things like this would also work

```
[root@fedora ~]# cp /bin/sleep /bin/sleep-ng
[root@fedora ~]# alternatives --install /bin/sleep-main sleep /bin/sleep 10
[root@fedora ~]# alternatives --install /sbin/sleep-main sleep /sbin/sleep-ng 10
the primary link for sleep must be /bin/sleep-main

```

P.S.: I hope that our mkosi conf for chkconfig actually works and I am really testing your code.

lnykryn commented:
Maybe something like:

```
int is_bin_or_sbin(const char *path, char **rest) {
        if (!path)
                return 0;

        if (strncmp(path, "/usr/bin/", strlen("/usr/bin/")) == 0) {
                *rest = (char *)path + strlen("/usr/bin/");
                return 1;
        } else if (strncmp(path, "/usr/sbin/", strlen("/usr/sbin/")) == 0) {
                *rest = (char *)path + strlen("/usr/sbin/");
                return 1;
        } else if (strncmp(path, "/bin/", strlen("/bin/")) == 0) {
                *rest = (char *)path + strlen("/bin/");
                return 1;
        } else if (strncmp(path, "/sbin/", strlen("/sbin/")) == 0) {
                *rest = (char *)path + strlen("/sbin/");
                return 1;
        }

        return 0;
}

int streq_sbin_merge(const char *a, const char *b) {
    if (streq(a, b))
        return 1;

    if (MERGED_SBIN && a && b) {
        char *rest_a = NULL;
        char *rest_b = NULL;
        if (is_bin_or_sbin(a, &rest_a) && is_bin_or_sbin(b, &rest_b))
            return streq(rest_a, rest_b);
    }

    return 0;

}

```

keszybz commented:
Updated with your change. I also renamed the function.

> I hope that our mkosi conf for chkconfig actually works and I am really testing your code.

Yes, it seems to work. But one needs to set MERGED_SBIN in the Makefile invocation to actually have any difference in behaviour. Once I do that, your example passes.
